modal/runner/engines/vllm.py

```diff
@@ -70,84 +84,82 @@ def __init__(self, params: VllmParams):
                 payload.prompt, params, payload.id
             )
 
-            final_output = None
+            output = ""
+            index = 0
             finish_reason = None
-            if payload.stream:
-                index = 0
-
-                async for request_output in results_generator:
-                    # Skipping invalid UTF8 tokens:
-                    if (
-                        request_output.outputs[0].text
-                        and request_output.outputs[0].text[-1] == "\ufffd"
-                    ):
-                        continue
-
-                    final_output = request_output
-                    token = final_output.outputs[0].text[index:]
-                    index = len(final_output.outputs[0].text)
-                    finish_reason = final_output.outputs[0].finish_reason
-                    yield create_sse_data(
-                        token,
-                        prompt_tokens=len(final_output.prompt_token_ids),
-                        completion_tokens=len(
-                            final_output.outputs[0].token_ids
-                        ),
-                        done=False,
-                        finish_reason=finish_reason,
-                    )
-
-                output = ""
-            else:
-                async for request_output in results_generator:
-                    final_output = request_output
-                    yield " "
-
-                output = final_output.outputs[0].text
-                finish_reason = final_output.outputs[0].finish_reason
-
-            prompt_tokens = len(final_output.prompt_token_ids)
-            completion_tokens = len(final_output.outputs[0].token_ids)
-            if payload.stream:
-                yield create_sse_data(
-                    "",
-                    prompt_tokens,
-                    completion_tokens,
-                    done=True,
-                    finish_reason=finish_reason,
-                )
-            else:
-                yield create_response_text(
-                    output,
-                    prompt_tokens,
-                    completion_tokens,
-                    done=True,
-                    finish_reason=finish_reason,
-                )
-
-            elapsed = time.perf_counter() - t_start_inference
+            async for current in results_generator:
+                output = current.outputs[0].text
+                finish_reason = current.outputs[0].finish_reason
+                resp.usage.prompt_tokens = len(current.prompt_token_ids)
+                resp.usage.completion_tokens = len(current.outputs[0].token_ids)
+                resp.usage.duration = time.time() - t_start_usage_duration
+
+                # Non-streaming requests continue generating w/o yielding intermediate results
+                if not payload.stream:
+                    yield " "  # QUESTION[sam]: why yield a single space for non-streaming?
+                    continue
```

Comment:
one more question here regarding the existing logic:

safe to remove the yielding empty spaces? or need them for a specific reason?

Reply:
removed in ba19e75 & it still seems to work fine on modal dev

Reply:
@sambarnes we need to yield "something" to keep the connection going, based on an experiment I did some time ago with fetch + SSE. Basically a hack to keep the connection going since this is a one-way pull, not a bi-directional connection like websocket where there's a dedicated handshake before transmission.

Reply:
Oh and the reason it's an empty space is so that the caller can parse the final JSON, since JSON trims spaces

Reply:
sick, makes sense! added it back & documented that reason
[a7f8a60](https://github.com/OpenRouterTeam/openrouter-runner/pull/68/commits/a7f8a6072f79b986edd565f77c485229b0c4494c)